src/PubSub/ResourceNameTrait.php

```diff
@@ -0,0 +1,144 @@
+<?php
+/**
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\PubSub;
+
+use InvalidArgumentException;
+
+trait ResourceNameTrait
+{
+    /**
+     * @var array
+     */
+    private $templates = [
+        'project' => 'projects/%1$s',
+        'topic' => 'projects/%2$s/topics/%1$s',
+        'subscription' => 'projects/%2$s/subscriptions/%1$s'
+    ];
+
+    /**
+     * @var array
+     */
+    private $regexes = [
+        'project' => '/^projects\/([^\/]*)$/',
+        'topic' => '/projects\/[^\/]*\/topics\/(.*)/',
```

Comment:
Hi Takashi,

To clarify: Your suggestion is to make the regex more precise so that is exactly matches the constraints described in the documentation? I can definitely do that. I just took these regular expressions from the [https://www.googleapis.com/discovery/v1/apis/pubsub/v1/rest](json schema). (The first occurrence is on line 469), so I'm certainly open to implementing a better regex if we decide that validating up front is useful. Maybe @dwsupplee could weigh in on whether that's the route we should take.

That being said, the regexes are only used to convert a full resource into a simple name. (i.e. `project/my-project/topics/my-topic` -> `my-topic`), and isn't used in the wrapper. It's provided mostly to add convenience to a user who might want the simpler name. There currently isn't validation on names coming in -- the API call will just fail if the resource name is malformed. [https://github.com/GoogleCloudPlatform/gcloud-node/blob/master/lib/pubsub/subscription.js#L344-351](gcloud-node) uses a similar strategy of leaving it to the API to fail if there's a problem.